core/runtime/src/main/java/io/quarkus/runtime/serial/Deserializer.java

```diff
@@ -0,0 +1,75 @@
+package io.quarkus.runtime.serial;
+
+import java.io.BufferedInputStream;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectStreamClass;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A deserialization utility.
+ */
+public final class Deserializer {
+    private Deserializer() {
+    }
+
+    private static final Map<String, Class<?>> primitives;
+
+    static {
+        Map<String, Class<?>> map = new HashMap<>();
+        map.put("boolean", boolean.class);
+
+        map.put("byte", byte.class);
+        map.put("short", short.class);
+        map.put("int", int.class);
+        map.put("long", long.class);
+
+        map.put("float", float.class);
+        map.put("double", double.class);
+
+        map.put("char", char.class);
+
+        map.put("void", void.class);
+        primitives = map;
+    }
+
+    /**
+     * Deserialize the given resource, throwing an error if it cannot be deserialized.
+     *
+     * @param type the result type class (must not be {@code null})
+     * @param name the resource to deserialize (must not be {@code null})
+     * @param <T> the result type
+     * @return the deserialized result
+     */
+    public static <T> T deserializeResource(Class<T> type, String name) {
+        final ClassLoader classLoader = Deserializer.class.getClassLoader();
+        return deserializeResource(type, name, classLoader);
+    }
+
+    static <T> T deserializeResource(final Class<T> type, final String name, final ClassLoader classLoader) {
+        final URL url = classLoader == null ? ClassLoader.getSystemResource(name) : classLoader.getResource(name);
+        if (url == null) {
+            throw new Error("Cannot deserialize due to missing " + name);
+        }
+        try (InputStream stream = url.openStream()) {
+            try (BufferedInputStream bis = new BufferedInputStream(stream)) {
+                try (ObjectInputStream ois = new ObjectInputStream(bis) {
+                    protected Class<?> resolveClass(final ObjectStreamClass desc) throws ClassNotFoundException {
```

Comment:
I don't have an opinion about the `ExecutionHandler` approach, but about this specific class, please be aware that any deserialization from an untrusted source is a security risk.

* https://christian-schneider.net/JavaDeserializationSecurityFAQ.html
* https://docs.oracle.com/en/java/javase/12/core/serialization-filtering1.html#GUID-3ECB288D-E5BD-4412-892F-E9BB11D4C98A
* https://support.symantec.com/us/en/article.symsa1344.html

Over at the Log4j2 project we deprecated the SerializationLayout that used to be the default in the SocketAppender and replaced it with JsonLayout because a security vulnerability was reported against the library that referred to this layout. Even whitelisting and blacklisting may leave security holes.

Generally, avoid deserialization to save yourself many headaches.

Reply:
> Generally, avoid deserialization to save yourself many headaches.

Thanks for the basic education on serialization. :)

But you should know that the latest version does not in fact use serialization anymore.

Reply:
Oh, sorry, I missed that. I was looking at the [Files changed](https://github.com/quarkusio/quarkus/pull/2866/files) link and I saw the Deserializer class. So this is no longer called and could be removed?

Reply:
Correct.